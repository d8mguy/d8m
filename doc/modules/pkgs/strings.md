## Strings

This module provides access to a subset of the functions in the corresponding go package, plus a few others.


#### function split

    val split = \(strg, sep: string) -> list(string) {...}
Be careful: go `Split` doesn't return an empty string if no instances of `sep` in `strg`, it returns a singleton list of strg. Thus, the correct test for no separators is split.count == 1, not 0.

#### function join

    val join = \(strgs: list(string), inter: string) -> string {...}


#### function toLower

    val toLower = \(strg: string) -> string {...}

#### function toUpper

    val toUpper = \(strg: string) -> string {...}

#### function toCap

    val toCap = \(strg: string) -> string {...}

#### function toCamel

    val toCamel = \(strgs: list(string), initial: boolean) -> string {...}
Generate a camel case string from the input strgs; `initial` says whether to capitalize the first word.

#### function genUpper

    val genUpper = \imp(inx: integer) -> string {...}
`Inx` encodes a base-26 number; generate an upper case letter sequence from it.

#### function genLower

    val genLower = \imp(inx: integer) -> string {...}
`Inx` encodes a base-26 number; generate a lower case letter sequence from it.

#### function toOrdinal

    val toOrdinal = \imp(xxx: integer) -> string {...}
This isn't the only way to generate ordinals but it's reasonable.

#### function startsWith

    val startsWith = \(strg, ptn: string) -> boolean {...}
True if `strg` starts with `ptn`.

#### function endsWith

    val endsWith = \(strg, ptn: string) -> boolean {...}
True if `strg` ends with `ptn`.

#### function trimSpace

    val trimSpace = \imp(strg: string) -> string {...}
Trim whitespace from both ends. This is done in a go pkg so we use that.

