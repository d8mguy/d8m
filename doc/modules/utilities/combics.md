## Combics

Combics provides a selection of useful combinatorial algorithms. The idea is to add more definitions over time.

#### enumCombns:function

    val enumCombns : \(n, k, totcount: integer) -> list(list(integer))

`enumCombns` returns a maximum of `totcount` of the combinations of n choose k. For this to make sense, we need `k >= 1` and `k < n`; `enumCombns` panics unless these conditions are met. Each combination is a list(integer) of length k of integers between 0 and n (exclusive); in addition, each combination will have strictly increasing values and the list of combinations is ordered lexicographically.

#### randperm:function

    val randperm : \mod(list(T)) -> list(T) for any T::entity

Randomly permute a list of type `T`, using go package `math/rand` with implicit seed to generate the random numbers. It modifies and returns the list. Note that repeated calls to this function will not generate all permutations of the list without repetition. Note also that since it uses the PRNG state built into `math/rand`, using this may change the behavior of other program elements that also use that PRNG. If this is a problem, you can easily write
a version with its own state; the code is just 8 lines long. The algorithm used here is called the Fisher-Yates shuffle; as explained in the wikipedia, a version was published in 1938 by Fisher and Yates, and popularized in vol 2 of Knuth's The Art of Computer Programming.

#### streamIndices:type

    val streamIndices = extend tuple(cur, limits: list(integer)) with {
        method lithook : \mod(lims: list(integer))
        method eltcount : \() -> integer
        method nDimensions : \() -> integer
        method restart : \mod() -> nothing
        method done = \() -> boolean
        method current = \() -> list(integer)
        method next = \mod() -> list(integer)
    }

A stream type that accepts a list of lengths and generates all the indices up to each of those lengths, in
lexicographic order. Thus, given [1,2,3] it will generate [[0,0,0],[0,0,1],[0,0,2],[0,1,0],[0,1,1],[0,1,2]].
Note that `[streamIndices: ]` is immediately done, it generates nothing. This is the right way to handle the situation where you need to enumerate all the elements of a variable number of lists. It's sensible to think of the k
values that initialize a `streamIndices` as the extents of a k-dimensional cube. From this perspective, the stream generates all vertices of the cube in lexicographic order. A  `streamIndices` will panic if any initializing length
is <= 0 or `next` is called when `done`.

A `stream type` by convention has methods named `current`, `next`, and `done`; you call `next` until `done` becomes true. 