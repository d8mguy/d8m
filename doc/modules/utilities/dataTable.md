## DataTable

Module dataTable lets you read 2d data tables in files or other generators, and do cube projections on the 2d data in memory. The module treats 2d data as `list(list(string))`, in other words data is uninterpreted except for tabular structure. Arbitrary dimensions can be held in a single data table, and related data tables can hold data that add more dimensions. 

A `paramT` defines a dimension for data. So files hold one or more dimensions in their output column and may or may not list the values of the parameters in the files themselves. All this is described in a `fileFormat`, which is held in a `datascheme`. The `fileFormat` describes the format of a single file. Besides the `fileFormat`, a `datascheme` holds a list of "xparams" and a function called dataGetter. The xparams describe the dimensions that are split across files, one element per file. The caller is responsible for assigning how filenames map to these elements, datascheme calls the dataGetter with the filenames supplied by the user and caches the contents. The name "dataGetter" is meant to emphasize that the "filename" can be any sort of descriptor; the source of data could be a URI/URL or anything convenient.

Besides the relevant types and their methods, dataTable exports stdFileGetter which you can use as a dataGetter for local files having a particularly simple format.

    val rowcolT = extend list(list(string)) where {
        method extractParameters = \imp(nparams: integer) -> rowcolT {...}
        method checkParameterEncoding = \imp(nparams: integer) -> string {...}
        method checkCounts = \(filename:string, xR, xC: integer) -> string {...}
    }
Method `extractParameters`: given a number of initial columns that are values for row encoded param, assumed in lexicographic order (column 0 varying slowest), return a `list(list(string))` with outer list indexed by param and inner lists being extracted lists of values. Code assumes nparams > 0.

Method `checkParameterEncoding`: do the same thing as `extractParameters` but check consistency instead of extracting element values. The consistency checking that's done is to verify that the values in the first N columns (N=# of params) repeat in the way dictated by enumerating those parameters lexicographically. Return a string describing the first problem found in the encoding, "" if no problem found. This works a lot harder than extractParameters; call it if you have any doubts about the integrity of your data source.

Method `checkCounts`: this is used when reading a file into the cache. It returns a non-empty string if there is a problem.


    val paramT = tuple(ident, descrn, unit: string, elts: list(string), rowEncoded:boolean, explicit: boolean)
`ParamT` describes the parameters encoded in a file: `ident` suitable as a label and `descrn` as a longer explanation; `unit` optional, `elts` enumerates the values this parameter takes in the file. The two booleans describe whether and how this parameter occurs in the file: if explicit then must be row encoded; if implicit can be encoded in either row or cols (the latter if `rowEncoded` is false). Another constraint on these encodings: all implicit column encoded params must occur after row encoded ones. All these constraints are checked in the `fileFormat` constructor.

    val stdFileGetter = \imp(fnm: string) -> rowcolT {...}
`DataScheme` has a `dataGetter:\(string)->rowcolT` attribute; `stdFileGetter` is a simple example of a suitable function. Given a path to a csv file with lines split on `\n`, no quotation, extra spaces, or extra lines, open, read, and close the file, split and return contents.

    [fielFormat: prms, outs]
where `prms: list(paramT)`, `outs: list(string)`. 

    val fileFormat = extend tuple(params: list(paramT), outputs: list(string)) where {
        method colEncodedCount = \() -> integer {...}
        method firstOutputIndex = \() -> integer {...}
        method nCols = \() -> integer {...}
        method nRows = \() -> integer {...}
    }
A file has `R` rows and `C` cols where `C = params.count + outputs.count`, `R = reduce(params.{elts.count}, $*, 1)`.

    [datascheme: ffmt, dg]
where `ffmt: fileFormat` and `dg:\(string)->rowcolT`

    val datascheme = extend tuple(filefmt: fileFormat, private xtparams: list(paramT), 
        dataGetter: \(string)->rowcolT) where {
        method update = \mod(xtprms:list(paramT), fnms:list(string), fpfx: string) -> nothing {...}
        method extract = \imp(pIndices, outputs: list(integer), colThresh: integer) -> rowcolT {...}
    }
A `datascheme` allows for extending a `fileFormat` across multiple "files" related by "xtparams" (external params); the `dataGetter` parameterizes the process of retrieving a data table. Internal to `datascheme` we manage a caching scheme so that the dataGetter is called at most once for each distinct filename. Note that xtparams is private to encourage you to change it only via `update`. Also, that an earlier design contemplated using a function/streamer to generate the correct sequence of filenames to pass to dataGetter but that in the end it was easier to provide a list of filenames.

Method `update`: changes xtparams and corresponding file related things.

Method `extract`: produces a data table of strings from a spec for extraction from self. The spec comes from 3 args: pIndices is a list of integers where >= 0 means that index in the corresponding param or xtparam and < 0 means the corresponding param or xtparam is enumerated in the output. Negative numbers should all be distinct and contiguous; their absolute magnitude sets the order of enumerated params in the output, aka "multis". Outputs tell which outputs of the scheme are produced. The numbers in this argument index the outputs slot of `filefmt.outputs`. To get the columns of the file where the data lives, we need to add the # of params, since those are listed in the first cols. Furthermore, `extract` can encode param values in cols of the output. The `colThresh` tells where to do this. It is a positive integer, we require `colThresh` <= the number of multis. For example, given pIndices = [1, -1, -3, 4, -2] and colThresh = 2. There are a total of 5 params+xtparams. We extract three of these, namely the ones at index 1, 2, and 4 in the concatentation of params and xtparams. In the output, we order the extracted dimensions with 1 in rows, 2 and 4 in cols, with 2 innermost. So if, for example, there are two outputs], param 2 has 3 elements and param 4 has 2 elements, then there will be 2*3*2 output columns. The order will be o02040, o12040, o02140, o12140, o02240, o12240, o02041, o12041, etc. The number of rows will equal the number of elts in params 1. Whether multiple files are involved depends on where the boundary of params and xtparams is. For example, if there are 3 params, then the -2 in pIndices concerns the 2nd xparam. We just assumed it has 2 elts so we'll be cycling between 2 files as we generate cols.