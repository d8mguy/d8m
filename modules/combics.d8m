// Combics provides a selection of useful combinatorial algorithms. This is intended to be an evolving module for quite some time.

export randperm, enumCombns, streamIndices

// Return a list(list(integer)) where each element is a length k list chosen from 0...n.
val enumCombns = \imp(n, k, totcount: integer) {
    if(k < 1 || k >= n) exit("bad args")
    var proto = (1..k).{index}
    var rslt: list(list(integer)) = [copy(proto)]
    val nkdiff = n - k
    var incrAt = k - 1
    // while condition is final exit criterion
    while(incrAt > 0 || proto[0] < n - k) {
        // loop through the options at current incrementing position
        while((incrAt < k - 1 && proto[incrAt] < proto[incrAt+1]+1) || proto[incrAt] < n - 1) {
            proto[incrAt] += 1
            rslt.pushb(copy(proto))
        }
        if(rslt.count > totcount) break
        // then reset incrAt and values in proto to set up next incrementing sequence
        while(incrAt > 0) {
            incrAt -= 1
            if(proto[incrAt] < nkdiff + incrAt) {
                var nval = proto[incrAt]+1
                each(h^incrAt...k) {
                    proto[h] = nval
                    nval += 1
                }
                rslt.pushb(copy(proto))
                incrAt = k - 1
                break
            }
        }
    }
    rslt
}

// A stream type that accepts a list of lengths and generates all the indices up to each of those lengths, in
// lexicographic order. Thus, given [1,2,3] it will generate [[0,0,0],[0,0,1],[0,0,2],[0,1,0],[0,1,1],[0,1,2]].
// Note that `[streamIndices: ]` is immediately done, it generates nothing. This is the right way to handle the
// situation where you need to enumerate all the elements of a variable number of lists. It's sensible to think of the k
// values that initialize a `streamIndices` as the extents of a k-dimensional cube. From this perspective, the
// stream generates all vertices of the cube in lexicographic order. A  `streamIndices` will panic if any initializing length
// is <= 0 or `next` is called when `done`.
//
// A `stream type` by convention has methods named `current`, `next`, and `done`; you call `next` until `done` becomes true.
val streamIndices = extend tuple(cur, limits: list(integer)) where {
    method lithook = \mod(lims: list(integer)) {
        if(any(lims.{this < 1})) exit("streamIndices requires positive limits")
        limits = lims.{this - 1}
        cur = zerolist(lims.count)
        unless(lims.count == 0) cur[lims.count - 1] = -1     // initialize so next returns all zeros first time
    }
    // utility method: number of elts produced by self from create to done
    method eltcount = \() { reduce(limits, \(a,b: integer) { a*(b+1)}, 1) }
    // utility method: number of "dimensions"
    method nDimensions = \() { cur.count }
    method restart = \mod() {
        cur = zerolist(limits.count)
        unless(limits.count == 0) cur[limits.count - 1] = -1
    }
    method done = \() { cur == limits }
    method current = \() { cur }
    method next = \mod() -> list(integer) {
        val nelts = limits.count
        each(i^1..nelts) {
            val ii = nelts - i; if(cur[ii] < limits[ii]) { cur[ii] += 1; each(j^(ii + 1)...nelts) cur[j] = 0; return cur }
        }
        exit("should not call terminated stream")
    }
}

// Randomly permute a list(T), using go package `math/rand` with implicit seed to generate the random numbers. It modifies and returns the list.
// Note that repeated calls to this function will not generate all its permutations without repetition.
// Note also that since it uses the PRNG state built into `math/rand`, using this may change
// the behavior of other program elements that also use that PRNG. If this is a problem, you can easily write
// a version with its own state; the code is just 8 lines long.
// The algorithm here is called the Fisher-Yates shuffle; as explained in the wikipedia, a version was
// published in 1938 by Fisher and Yates, and popularized in vol 2 of Knuth's TAOCP.

given(T::entity) {
    val randperm = \mod(lst: list(T), randSrc: \(integer)->integer) {
        var lim = count(lst)
        each(i^0...count(lst)-1) {
            val j = i+randSrc(lim)
            swap(lst, i, j)
            lim -= 1
        }
        lst
    }
}
