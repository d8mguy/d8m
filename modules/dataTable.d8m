// Module dataTable lets you manipulate 2d data tables in files or other generators, and do cube projections on the 2d data in memory.
// This version treats 2d data as `list(list(string))`, in other words data is uninterpreted.
// Arbitrary dimensions can be held in a single data table, and related data tables can hold data for more dimensions.
// A `paramT` defines a dimension for data. So files hold one or more dimensions in their output column and may or may not
// list the values of the parameters in the files themselves. All this is described in a `fileFormat`, which is held in a
// `datascheme`. The `fileFormat` describes the format of a single file.
    // Besides the `fileFormat`, a `datascheme` holds a list of "xparams" and a function called dataGetter. The
// xparams describe the dimensions that are split across files, one element per file. The caller is responsible for assigning
// how filenames map to these elements, datascheme calls the dataGetter with the filenames supplied by the user and caches
// the contents. The name "dataGetter" is meant to emphasize that the "filename" can be any sort of descriptor; the source
// of data could be a URI/URL or anything convenient.
//
// Besides the relevant types and their methods, dataTable exports `stdFileGetter` which you can use as a dataGetter for
// local files having a particularly simple format.


import "combics"
import "file"
import "bytes"
import "strings"


export datascheme, paramT, fileFormat, stdFileGetter, rowcolT

// A rowcolT is just a list(list(string)) with a couple of useful methods. The outer list holds rows, so cols are inner.
val rowcolT = extend list(list(string)) where {
    // Given a number of initial columns that are values for row encoded param, assumed in lexicographic order (column 0 varying slowest), return
    // a list(list(string)) with outer list indexed by param and inner lists being extracted lists of values.
    // Code assumes nparams > 0
    method extractParameters = \imp(nparams: integer) -> rowcolT {
        if(nparams == 1) return [self.{ this[0] }]
        var rslt: rowcolT = []
        var curprm = nparams - 2, stride = 1
        while(curprm >= 0) {
            val c0 = self[0][curprm]
            var rpts = 0, inx = 0, elts: list(string) = []
            while(self[inx][curprm] == c0) { rpts += 1; elts.pushb(self[inx][curprm+1]); inx += stride }
            rslt.pushb(elts)
            stride *= rpts
            curprm -= 1
        }
        rslt
    }

    // Do the same thing as extractParameters but check consistency instead of extracting element values.
    // The consistency checking that's done is to verify that the values in the first N cols (N=# of params) repeat in
    // the way dictated by enumerating those parameters lexicographically.
    // Return a string describing the first problem found in the encoding, "" if no problem found.
    // This works a lot harder than extractParameters; call it if you have any doubts about the integrity of your data source.
    method checkParameterEncoding = \imp(nparams: integer) -> string {
        // Note that there's nothing to detect if a single param
        if(nparams <= 1) return ""
        var rslt: rowcolT = []
        var curprm = nparams - 2, stride = 1
        // work bkwds through the params, record values and ensure they repeat at the right places.
        while(curprm >= 0) {
            val c0 = self[0][curprm]
            var rpts = 0, inx = 0, elts: list(string) = []
            while(self[inx][curprm] == c0) { rpts += 1; elts.pushb(self[inx][curprm+1]); inx += stride }
            rslt.pushb(elts)
            // check the rest of the rows to verify they have the same pattern of repeats
            while(inx < self.count) {
                each(elt^elts, inx0) unless(self[inx + inx0*stride][curprm+1] == elt) return "match fail in row #{inx + inx0*stride}, col #{curprm+1}"
                inx += rpts * stride
            }
            // todo: consider if I need a check on #rows
            stride *= rpts
            curprm -= 1
        }
        ""
    }

    // This is used when reading a file into the cache. It returns a non-empty string if there is a problem.
    method checkCounts = \(filename:string, xR, xC: integer) -> string {
        val colcounts = self.{this.count}.uniq
        if(self.count == 0) "file #{filename} is empty"
        else if(self.count == xR && colcounts.count == 1 && colcounts[0] == xC) ""
        else "format of #{filename}: expected #{xR}*#{xC} but got #{self.count}*#{colcounts[0]}"
    }

}

// describes the parameters encoded in a file: ident suitable as a label and descrn as a longer explanation; unit optional,
// elts enumerates the values this parameter takes in the file. The two booleans describe whether and how this parameter
// occurs in the file: if explicit then must be row encoded; if implicit can be encoded in either row or cols (the latter
// if rowEncoded is false). Another constraint on these encodings: all implicit column encoded params must occur after row
// encoded ones. All these constraints are checked in the fileFormat constructor.
val paramT = tuple(ident, descrn, unit: string, elts: list(string), rowEncoded:boolean, explicit: boolean)

// Datascheme has a dataGetter:\(string)->rowcolT attribute; stdFileGetter is a simple example of a suitable function.
// Given a path to a csv file with lines split on \n, no quotation, extra spaces, or extra lines, [open, read, close]
// the file, split and return contents.
val stdFileGetter = \imp(fnm: string) -> rowcolT {
    var fd = file.openBasic(fnm)
    val buff: list(byte) = zerolist(fd.size())
    if(buff.count == 0) exit("can't handle empty file")
    fd.read(buff)
    fd.close()
    val rows = bytes.split(buff, ['\n'])
    var splitrows = rows.{ bytes.split(this, ",").{ cvt(this, string) } }
    // The bytes.Split ftn is pretty lame with a final newline, which is pretty common, it creates an extra line with an empty string
    // This is tricky to get rid of, so I'm taking a shortcut
    if(buff[buff.count-1] == '\n') splitrows = splitrows[0...splitrows.count - 1]
    cast(splitrows, rowcolT)
}

// A file has R rows and C cols where C = params.count + outputs.count, R = reduce(params.{elts.count}, $*, 1).
val fileFormat = extend tuple(params: list(paramT), outputs: list(string)) where {
    method lithook = \mod(prms: list(paramT), outs: list(string)) {
        if(any(prms.{explicit && !this.rowEncoded})) exit("file format has paramT's with illegal combination of booleans")
        val colEnc0 = prms[!this.rowEncoded => index]
        if(colEnc0 != nil && prms[colEnc0...prms.count][this.rowEncoded] != nil) exit("row encoded parameters may not occur after column encoded ones")
        params = prms
        outputs = outs
    }
    // This will be the number of columns needed to record the values of all column encoded params.
    method colEncodedCount = \() -> integer {
        val colEnc0 = params[!this.rowEncoded => index]
        outputs.count * (colEnc0 == nil ? 1 : reduce(params.[index >= colEnc0].{elts.count}, $*, 1))
    }
    method firstOutputIndex = \() -> integer { params.[explicit].count }
    method nCols = \() -> integer { colEncodedCount() + firstOutputIndex() }
    method nRows = \() -> integer { reduce(params.[rowEncoded].{ elts.count }, $*, 1) }
}

// A `datascheme` allows for extending a `fileFormat` across multiple "files" related by "xtparams" (external params); the
// `dataGetter` parameterizes the process of retrieving a data table. Internal to `datascheme` we manage a caching scheme so
// that the dataGetter is called at most once for each distinct filename.
// Note that xtparams is private to encourage you to change it only via `update`. Also, that an earlier design contemplated
// using a function/streamer to generate the correct sequence of filenames to pass to dataGetter but that in the end it
// was easier to provide a list of filenames.
val datascheme = extend tuple(filefmt: fileFormat, private xtparams: list(paramT), dataGetter: \(string)->rowcolT) where {
    private attribute filenames: list(string)
    private attribute fnPrefix: string
    private attribute cache: list(nilPossible(rowcolT))

    // Construct without xtparams then use update method to add xtparams, etc.
    method lithook = \mod(ffmt: fileFormat, dg: \(string)->rowcolT) { filefmt = ffmt; dataGetter = dg }

    method xparamsCount = \() { xtparams.count }

    // Provides unified access to elt counts for params and xtparams. Sanity checking must be done before calling this.
    private method allParamsEltCounts = \(pInx: integer) {
        val nParams = filefmt.params.count
        if(pInx < nParams) filefmt.params[pInx].elts.count else xtparams[pInx - nParams].elts.count
    }

    // Changes xtparams and corresponding file related things.
    method update = \mod(xtprms:list(paramT), fnms:list(string), fpfx: string) -> nothing {
        xtparams = xtprms
        filenames = fnms
        fnPrefix = fpfx
//        println("update sets filenames to ", fnms)
        // it would be possible to try to save cached items with the same filename but it's simpler (& less code) to start over
        cache = (1..fnms.count).{ cast(nil, nilPossible(rowcolT)) }
    }
    assert noInline(update)

    // Returns the string value of the data point indicated by the args, handling cache misses as needed.
    // Todo: extend this to handle reading col encoded params.
    private method datum = \mod(file, row, col, xpectRows, xpectCols: integer) {
        var fcontents = cache[file]
        if(fcontents == nil) {
            val fnm = filenames[file]
            fcontents = dataGetter(fnPrefix + fnm)
            assert fcontents != nil
            val errmsg = fcontents.checkCounts(fnm, xpectRows, xpectCols)
            if(errmsg != "") exit(errmsg)
            cache[file] = fcontents
        }
        fcontents[row][col]
    }
    assert purified(datum)

    // Produces a data table of strings from a spec for extraction from self. The spec comes from 3 args: `pIndices` is
    // a list of integers where >= 0 means that index in the corresponding param or xtparam and < 0 means the corresponding
    // param or xtparam is enumerated
    // in the output. Negative numbers should all be distinct and contiguous; their absolute magnitude sets the
    // order of enumerated params in the output, aka "multis". Outputs tell which outputs of the scheme are produced. The numbers in this
    // argument index the outputs slot of filefmt.outputs. To get the columns of the file where the data lives, we
    // need to add the # of params, since those are listed in the first cols.
    // Furthermore, extract can encode param values in cols of the output.
    // The colThresh tells where to do this. It is a positive integer, we require colThresh <= the number of multis.
    // For example, given pIndices = [1, -1, -3, 4, -2] and colThresh = 2. There are a total of 5 params+xtparams. We
    // extract three of these, namely the ones at index 1, 2, and 4 in the concatentation of params and xtparams. In the
    // output, we order the extracted dimensions with 1 in rows, 2 and 4 in cols, with 2 innermost. So if, for example,
    // there are two outputs], param 2 has 3 elements and param 4 has 2 elements, then there will be 2*3*2 output columns.
    // The order will be o02040, o12040, o02140, o12140, o02240, o12240, o02041, o12041, etc. The number of rows will equal
    // the number of elts in params 1. Whether multiple files are involved depends on where the boundary of params and xtparams
    // is. For example, if there are 3 params, then the -2 in pIndices concerns the 2nd xparam. We just assumed it has 2 elts
    // so we'll be cycling between 2 files as we generate cols.
    method extract = \imp(pIndices, outputs: list(integer), colThresh: integer) -> rowcolT {
        //println("start extract:", filefmt.params, pIndices, outputs)
        // Multis are indices of concatenated params and xtparams dimensions: multis[0] = pIndices[this == -1 => index], etc.
        // So multis.count is the #multidims.
        var multis:list(integer) = zerolist(max(pIndices.{-this}, 0))
        var multisCounted = 0       // we'll vfy that enumd indices are contig by ensuring this count equals multis.count
        val nParams = filefmt.params.count, nXParams = xtparams.count
        // First, do some sanity checks, panic on any error
        unless(pIndices.count == nParams + nXParams) exit("wrong number of indices to extract (got #{pIndices.count} but expected #{nParams+nXParams}")
        each(pInx^pIndices, inxInIndices) {
            if(pInx < 0) { multis[-(pInx+1)] = inxInIndices; multisCounted += 1 }
            else if(inxInIndices < nParams) { unless(pInx < filefmt.params[inxInIndices].elts.count) exit("arg #{inxInIndices} binding out of range") }
            else { val nXpelts = xtparams[inxInIndices - nParams].elts.count; unless(pInx < nXpelts) exit("xp arg #{inxInIndices - nParams} binding out of range") }
        }
        if(colThresh > multisCounted) exit("extract: colThresh range error")
        if(multis.count != multisCounted) exit("extract: pIndices has missing or duplicated negative values")

        // The args are ok. Set up the various state machines that we'll run to generate output.
        // Separately for both params and xtparams, we get elt counts, calculate accumulated weights (need to reverse first
        // because of how accumulate works), then drop the total product, which isn't needed an lines up our indices correctly.
        // The increments of enumerated params come directly from the corresponding weight (of the params after it); the
        // initial offset, which is incremented modularly for each datapoint access, is the dot product of the weights with
        // the offsets of each param, enumerated params set to zero. Exactly the same logic applies to xtparams, except that
        // they concern fileIndex instead of rowOffset.
        val paramCounts = reverse(filefmt.params.{ this.elts.count })
        // Accumulate puts the total product at index 0, so we strip the first elt
        val paramWghts = reverse(paramCounts.accumulate($*, 1))
        val paramWghtsTail = paramWghts.[index >= 1]
        val offsetsVect = pIndices.[index < paramWghtsTail.count].{ max(this, 0) }     // copy, setting neg elts to 0
        var rowOffset = dotprod(paramWghtsTail, offsetsVect)
        val xtparamCounts = reverse(xtparams.{ this.elts.count })
        val xtparamWghtsTail = reverse(xtparamCounts.accumulate($*, 1)).[index >= 1]
        val xtOffsetsVect = pIndices.[index >= paramWghtsTail.count].{ max(this, 0) }
        var fileIndex = dotprod(xtOffsetsVect, xtparamWghtsTail)
//        println("orig xptw", xtparamCounts, xtparamWghtsTail, xtOffsetsVect, fileIndex)
//        println("multis:", multis)
//        println("xpt=", xtparams, "paramWghtsTail", paramWghtsTail, "rowOffset:", rowOffset, "xtpwt:", xtparamWghtsTail, "xtov=", xtOffsetsVect, "fi", fileIndex)
        // file enumeration arises from xtparams in multis
        val multiCounts = multis.{ allParamsEltCounts(this) - 1 }
        val outCols = outputs.{ this + filefmt.firstOutputIndex }
        // The xFsmT represents the FSM we need. We need to bump fileIndex (for xtparams) or rowOffset (for regular params)
        // by some fixed amount. So we code this in the incr slot: If incr > 0, add it to rowOffset, if < 0, bump
        // fileIndex by neg of amt. The list(xFsmT) is run in nested fashion: run through the last one, then incr next to last and
        // restart last, etc. Until first one is exhausted, at which point we're done. Same idea as combics.streamIndices
        // but instead of incrg an index we run a FSM.
        val xFsmT = extend tuple(cur, limit, incr, fcount: integer) where {
            method done = \() { cur >= limit }
            method restart = \mod() {
                cur = 0
                if(incr > 0) rowOffset -= limit*incr else { fileIndex += limit*incr; if(fileIndex < 0) fileIndex += fcount }
            }
            method next = \mod() {
                //println("next:", incr, rowOffset, fileIndex)
                if(incr > 0) rowOffset += incr else {
                    fileIndex -= incr
                    if(fileIndex >= fcount) fileIndex -= fcount
                    else if(fileIndex < 0) fileIndex += fcount
                }
                cur += 1
            }
        }
        var output: rowcolT = []
        var curcol: list(string) = []
        // We reverse the order for xtparams. I'm not sure this is right in all cases but it works for 1 substn+hmdim
        // Possible that the right answer is to leave the substns in order and put the hmdim first
//        println("multis=", multis, "multicounts=", multiCounts, "pwt=", paramWghtsTail, "xtpwt=", xtparamWghtsTail, "fi=", fileIndex, "ro=", rowOffset)
        var nest = multis.{ [xFsmT: 0, multiCounts[index], this >= nParams ? -xtparamWghtsTail[this - nParams] : paramWghtsTail[this], filenames.count] }
//        println("initial state", nest)
        var level = nest.count - 1
        val xpectRows = paramWghts[0]       // the xpect vbls are for sanity checking
        val xpectCols = nParams + filefmt.outputs.count
        loop {
            each(colinx^outCols) curcol.pushb(datum(fileIndex, rowOffset, colinx, xpectRows, xpectCols))
//            println("start incr, row=", rowOffset, "file=", fileIndex)
            while(level >= 0 && nest[level].done) level -= 1
            if(level < colThresh) {
                output.pushb(curcol)
//                println("push", curcol)
                curcol = []
            }
            if(level < 0) break
            nest[level].next
//            println("set ", level, "to", nest[level])
            while(level < nest.count - 1) {
                level += 1
                nest[level].restart
//                println("restart", level)
            }
        }
//        println("finish extract")
        output
    }

}
