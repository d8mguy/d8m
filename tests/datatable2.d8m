// try a different datascheme from test 1
import "dataTable" as csv
import "strings"
import "combics" melted

val params = [
    [csv.paramT: "P1", "", "", ["e00", "e01", "e02", "e03", "e04", "e05", "e06", "e07"], true, true]]
val xParams = [
    [csv.paramT: "X1", "", "", ["x00", "x01", "x02"], true, true]]
val outputs = ["o1", "o2", "o3", "o4"]
val synthDatatable1 = \imp(fnm: string) -> csv.rowcolT {
    var rownum = 0
    var pStrm = [streamIndices: params.{ elts.count }]
    var rslt: csv.rowcolT = []
    while(!pStrm.done) {
        var prms = pStrm.next.{ params[index].elts[this] }
        val outs = outputs.{ strings.join([this, stringify(rownum), fnm], "/") }
        rslt.pushb(prms.append(outs))
        rownum += 1
    }
    rslt
}
val foo = \imp() {
    var state = [csv.datascheme: [csv.fileFormat: params, outputs], synthDatatable1]
    val fnames = (0..2).{ "x0#{this}" }
    state.update(xParams, fnames, "")
    val rc0 = state.extract([-1, -2], [0], 1)
    println("state.extract([-1, -2], [0], 1):", rc0)
    val rc1 = state.extract([-2, -1], [0], 1)
    println("state.extract([-2, -1], [0], 1):", rc1)
    val rc2 = state.extract([-1, -2], [0], 2)
    println("state.extract([-1, -2], [0], 2):", rc2)
    val rc3 = state.extract([-2, -1], [0], 2)
    println("state.extract([-2, -1], [0], 2):", rc3)
}
foo()
