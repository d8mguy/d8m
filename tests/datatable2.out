state.extract([-1, -2], [0], 1): [[o1/0/x00, o1/0/x01, o1/0/x02], [o1/1/x00, o1/1/x01, o1/1/x02], [o1/2/x00, o1/2/x01, o1/2/x02], [o1/3/x00, o1/3/x01, o1/3/x02], [o1/4/x00, o1/4/x01, o1/4/x02], [o1/5/x00, o1/5/x01, o1/5/x02], [o1/6/x00, o1/6/x01, o1/6/x02], [o1/7/x00, o1/7/x01, o1/7/x02]]
state.extract([-2, -1], [0], 1): [[o1/0/x00, o1/1/x00, o1/2/x00, o1/3/x00, o1/4/x00, o1/5/x00, o1/6/x00, o1/7/x00], [o1/0/x01, o1/1/x01, o1/2/x01, o1/3/x01, o1/4/x01, o1/5/x01, o1/6/x01, o1/7/x01], [o1/0/x02, o1/1/x02, o1/2/x02, o1/3/x02, o1/4/x02, o1/5/x02, o1/6/x02, o1/7/x02]]
state.extract([-1, -2], [0], 2): [[o1/0/x00], [o1/0/x01], [o1/0/x02], [o1/1/x00], [o1/1/x01], [o1/1/x02], [o1/2/x00], [o1/2/x01], [o1/2/x02], [o1/3/x00], [o1/3/x01], [o1/3/x02], [o1/4/x00], [o1/4/x01], [o1/4/x02], [o1/5/x00], [o1/5/x01], [o1/5/x02], [o1/6/x00], [o1/6/x01], [o1/6/x02], [o1/7/x00], [o1/7/x01], [o1/7/x02]]
state.extract([-2, -1], [0], 2): [[o1/0/x00], [o1/1/x00], [o1/2/x00], [o1/3/x00], [o1/4/x00], [o1/5/x00], [o1/6/x00], [o1/7/x00], [o1/0/x01], [o1/1/x01], [o1/2/x01], [o1/3/x01], [o1/4/x01], [o1/5/x01], [o1/6/x01], [o1/7/x01], [o1/0/x02], [o1/1/x02], [o1/2/x02], [o1/3/x02], [o1/4/x02], [o1/5/x02], [o1/6/x02], [o1/7/x02]]
